framework/include/utils/FactorizedSymmetricRankTwoTensor.h

```diff
@@ -0,0 +1,300 @@
+
+  /// Returns _A rotated by R.
+  FactorizedSymmetricRankTwoTensorTempl<T> rotated(const RankTwoTensorTempl<T> & R) const;
+
+  /// Returns the transpose of _A.
+  FactorizedSymmetricRankTwoTensorTempl<T> transpose() const;
+
+  // @{ Assignment operators
+  FactorizedSymmetricRankTwoTensorTempl<T> &
+  operator=(const FactorizedSymmetricRankTwoTensorTempl<T> & A);
+  FactorizedSymmetricRankTwoTensorTempl<T> & operator=(const RankTwoTensorTempl<T> & A);
+  // @}
+
+  /// performs _A *= a, also updates eigen values
+  FactorizedSymmetricRankTwoTensorTempl<T> & operator*=(const T & a);
+
+  /// returns _A * a, also updates eigen values
+  template <typename T2, typename std::enable_if<ScalarTraits<T2>::value, int>::type = 0>
+  FactorizedSymmetricRankTwoTensorTempl<typename CompareTypes<T, T2>::supertype>
+  operator*(const T2 & a) const;
+
+  /// performs _A /= a, also updates eigen values
+  FactorizedSymmetricRankTwoTensorTempl<T> & operator/=(const T & a);
+
+  /// returns _A / a, also updates eigen values
+  template <typename T2, typename std::enable_if<ScalarTraits<T2>::value, int>::type = 0>
+  FactorizedSymmetricRankTwoTensorTempl<typename CompareTypes<T, T2>::supertype>
+  operator/(const T2 & a) const;
+
+  /// Defines logical equality with another RankTwoTensorTempl<T>
+  bool operator==(const FactorizedSymmetricRankTwoTensorTempl<T> & A) const;
+
+  /// inverse of _A
+  FactorizedSymmetricRankTwoTensorTempl<T> inverse() const;
+
+  /// add identity times a to _A
+  void addIa(const T & a);
+
+private:
+  // Assemble the tensor from the factorization
+  RankTwoTensorTempl<T> assemble() const
+  {
+    RankTwoTensorTempl<T> D(_eigvals);
+    return _eigvecs * D * _eigvecs.transpose();
+  }
+
+  // The eigen values of _A;
+  std::vector<T> _eigvals;
+
+  // The eigen vectors of _A;
+  RankTwoTensorTempl<T> _eigvecs;
+};
+
+namespace MathUtils
+{
+template <typename T>
+FactorizedSymmetricRankTwoTensorTempl<T>
+log(const FactorizedSymmetricRankTwoTensorTempl<T> & A)
+{
+  const auto & eigvals = A.eigvals();
+  return FactorizedSymmetricRankTwoTensorTempl<T>(
+      {std::log(eigvals[0]), std::log(eigvals[1]), std::log(eigvals[2])}, A.eigvecs());
+}
+
+template <typename T>
+FactorizedSymmetricRankTwoTensorTempl<T>
+exp(const FactorizedSymmetricRankTwoTensorTempl<T> & A)
+{
+  const auto & eigvals = A.eigvals();
+  return FactorizedSymmetricRankTwoTensorTempl<T>(
+      {std::exp(eigvals[0]), std::exp(eigvals[1]), std::exp(eigvals[2])}, A.eigvecs());
+}
+
+template <typename T, typename T2, typename std::enable_if<ScalarTraits<T2>::value, int>::type>
+FactorizedSymmetricRankTwoTensorTempl<typename CompareTypes<T, T2>::supertype>
+pow(const FactorizedSymmetricRankTwoTensorTempl<T> & A, const T2 & p)
+{
+  const auto & eigvals = A.eigvals();
+  return FactorizedSymmetricRankTwoTensorTempl<typename CompareTypes<T, T2>::supertype>(
+      {std::pow(eigvals[0], p), std::pow(eigvals[1], p), std::pow(eigvals[2], p)}, A.eigvecs());
```

Comment:
Is this only working for Internet powers? Or what does the scalar traits stuff mean?

Reply:
Fixed.